### PR TITLE
stylua: 0.14.3 -> 0.15.0

### PR DESCRIPTION
**Pull request by figsoda**
###### Description of changes

I don't know what the default feature set should be. The original repo has no default features, but I haven't been following the original. I also don't want to break backwards compatibility in nixpkgs, but 5.2 is kinda a weird version to stop at, so I changed to default to 5.4

https://github.com/JohnnyMorganz/StyLua/compare/v0.14.3...v0.15.0
https://github.com/JohnnyMorganz/StyLua/blob/main/CHANGELOG.md

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
> https://github.com/JohnnyMorganz/StyLua#with-github-releases
> By default, these are built with all syntax variants enabled (Lua 5.2, 5.3, 5.4 and Luau), to cover all possible codebases

Upstream's binaries are built with `all-features`, seems reasonable to follow that?

**Comment by figsoda** [comment]
I will turn on lua54 and luau then, not turning on lua52 and lua53 by default so it is easier to do e.g. lua51 only

**Comment by zowoq** [comment]
I'd rather we follow upstream binaries and do `all-features` or follow the crate default and have none.

**Comment by figsoda** [comment]
the lua54 feature implies lua53, and the same for lua52
I don't think something like `allFeatures ? true` fits here

**Comment by zowoq** [comment]
How about sometime like this so we don't need to pick and/or track specific features and people can still override it if they need to?

```diff
diff --git a/pkgs/development/tools/stylua/default.nix b/pkgs/development/tools/stylua/default.nix
index 492263f2c36..ee997a6839c 100644
--- a/pkgs/development/tools/stylua/default.nix
+++ b/pkgs/development/tools/stylua/default.nix
@@ -1,10 +1,7 @@
 { lib
 , rustPlatform
 , fetchFromGitHub
-, lua52Support ? false
-, lua53Support ? false
-, lua54Support ? true
-, luauSupport ? true
+, buildFeatures ? "all-features"
 }:
 
 rustPlatform.buildRustPackage rec {
@@ -25,10 +22,7 @@ rustPlatform.buildRustPackage rec {
     rm .cargo/config.toml
   '';
 
-  buildFeatures = lib.optional lua52Support "lua52"
-    ++ lib.optional lua53Support "lua53"
-    ++ lib.optional lua54Support "lua54"
-    ++ lib.optional luauSupport "luau";
+  inherit buildFeatures;
 
   meta = with lib; {
     description = "An opinionated Lua code formatter";

```

**Comment by figsoda** [comment]
I don't think all-features would have worked in this case so I just switched to `[ "lua54" "luau" ]`, also named it features instead of buildFeatures

**Comment by zowoq** [comment]
> I don't think all-features would have worked in this case 

Why not?

**Comment by figsoda** [comment]
> error: Package `stylua v0.15.0 (/build/source)` does not have the feature `all-features`

in ci it is taken as a flag (`--all-features`) and `--features all-features` doesn't work

**Comment by figsoda** [comment]
added a release notes entry since this breaks backwards compatibility

**Comment by zowoq** [comment]
> > error: Package `stylua v0.15.0 (/build/source)` does not have the feature `all-features`
> 
> in ci it is taken as a flag (`--all-features`) and `--features all-features` doesn't work

Set buildflags/checkflags to use `--all-flags` if features is unset?

**Comment by zowoq** [comment]
> > the lua54 feature implies lua53, and the same for lua52

Apart from this there a reason not to have 52 and 53 here? If we not going to actually use `all-features` I think we should set the equivalent as the default.

**Comment by figsoda** [comment]
because it is shorter and this is the equivalent, lua54 implies lua53 and lua52
https://github.com/JohnnyMorganz/StyLua/blob/v0.15.0/Cargo.toml#L29-L30

**Comment by zowoq** [comment]
> because it is shorter and this is the equivalent, lua54 implies lua53 and lua52

Yes, I was asking if that was the only reason. For clarity I'd rather we set all of the features.

File: pkgs/development/tools/stylua/default.nix [hunk 1]
```diff
@@ -1,34 +1,28 @@
-{ fetchFromGitHub
-, lib
+{ lib
 , rustPlatform
-, lua52Support ? true
-, luauSupport ? false
-, fetchpatch
+, fetchFromGitHub
+, features ? [ "lua54" "luau" ]
```
Comment (zowoq) [inline pkgs/development/tools/stylua/default.nix:4]: Guess we could just add a comment.
```suggestion
  # lua54 implies lua52/lua53
, features ? [ "lua54" "luau" ]
```

**Comment by figsoda** [comment]
it was the only reason, conditionally setting `buildFlags` just seemed kinda hacky to me when the current solution works well enough

**Pull request merged**